File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/coding_interviews_2/Question_0043_NumberOf1.java
package com.lanceolata.coding.algorithms.java.coding_interviews_2;

public class Question_0043_NumberOf1 {
    public int NumberOf1Between1AndN_Solution(int n) {
        int count = 0, high = 0, cur = 0, low = 0;
        for (int i = 1; i <= n; i *= 10) {
            high = n / (i * 10);
            cur = (n / i) % 10;
            low = n % i;
            if (cur == 0) {
                count += high * i;
            } else if (cur == 1) {
                count += high * i + low + 1;
            } else {
                count += (high + 1) * i;
            }
        }
        return count;
    }
}

File: algorithms/algorithms-cplusplus/coding_interviews_2/Question_0043_NumberOf1.cc

class Solution {
 public:
  int NumberOf1Between1AndN_Solution(int n) {
    int count = 0;
    long long i=1;
    int high = 0, cur = 0, low = 0;
    for(i = 1; i <= n; i *= 10) {
      high = n / (i * 10); // 高位
      cur = (n / i) % 10;  // 当前位
      low = n % i;         // 低位
      if (cur == 0) {
        count += high * i;
      } else if (cur == 1) {
        count += high * i + low + 1;
      } else {
        count += (high + 1) * i;
      }
    }
    return count;
  }
};
